### PR TITLE
feat: add prioritizing based on role and user hierarchy

### PR DESCRIPTION
**Pull request by tangyang9464**
Signed-off-by: tangyang9464 <tangyang9464@163.com>
Fix: #831
@hsluoyz @closetool plz review
@merajsahebdar plz just change ``priority(p.eft) || deny`` -> ``subjectPriority(p.eft) || deny``  and test your case.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@closetool plz review.

**Comment by tangyang9464** [comment]
For this feature, the inherited structure of roles and users can only be multiple trees, not graphs. One user cannot correspond to multiple roles.such as ``g, jane,editor`` ``g, jane,subscriberr`` ,the ``jane`` have two roles.This situation does not guarantee the correct result

**Comment by hsluoyz** [comment]
@merajsahebdar plz review.

**Comment by merajsahebdar** [comment]
@tangyang9464 
I tried a lot of scenarios, It's working great.
Thanks for your time and effort! 🙏🏻 Now Casbin becomes more lovely to me!

Just to make it a little more clear about what you mentioned, it works even  in a scenario like this:
```
p, root, data1, write, allow
p, admin, data1, write, deny
p, editor, data1, write, deny

p, user, data1, read, allow
p, subscriber, data1, read, deny

g, admin, root
g, editor, admin
g, subscriber, user

g, jane, root
g, jane, user
```
but does not work in:
```
p, root, data1, write, allow
p, admin, data1, write, deny
p, editor, data1, write, deny

p, user, data1, read, allow
p, user, data1, write, allow # trying to rewrite policies in tree: root
p, subscriber, data1, read, deny

g, admin, root
g, editor, admin
g, subscriber, user

g, jane, root
g, jane, user
``` 
so, users can have multiple roles, but:
1. those roles could not be in the same tree. (logically it's weird to add a user in two different branches of a tree, so might not happen!)
2. you could not define the same policy for both of those different types of roles.

However, I don't think these happen in a well-designed logic, but it's necessary  to make it clear in docs.

Thanks again!

**Comment by tangyang9464** [comment]
@merajsahebdar
Thanks for your advice！But I want to explain something.
1. In your example,``jane`` has the same hierarchy(level) in both trees,so the plolicy ranking will not be affected
2. But Consider the following example

model
```
p, user, data1, read, allow
p, subscriber, data1, read, deny

g, admin1, root1
g, admin2, root2

g, jane,admin1
g, jane,root2 # jane has two roles and the levels in two tree are different

g, alice,admin2
```
the corresponding tree structure
image
the level:

- ``jane``  in `root-1` tree is 3
- But ``jane`` in  ``root-2`` tree is 2

so sometimes ``admin-1`` is 2,but the ``jane`` is 2 also.it's wrong(each role or user can onle has a level.If has two,will overwrite each other)

if one user has multiple roles,You have to make sure the user has the same level in different trees and in your second point
``you could not define the same policy for both of those different types of roles.``
In this case,the order for ``root`` and ``user`` is the default（from top to bottom）because they level are same

I will modify the code so that can report errors when using incorrectly(such as one user have two level)

File: examples/subject_priority_policy.csv [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+p, jane, data1, read, allow
+p, alice, data1, read, allow
+
+g, admin, root
+
+g, editor,admin
+g, subscriber,admin
+
+g, jane,editor
+g, alice,subscriber
```
Comment (closetool) [inline examples/subject_priority_policy.csv:16]: Add blank after comma.

File: model/model.go [hunk 2]
```diff
@@ -206,6 +208,89 @@ func (model Model) PrintModel() {
...
+	for ptype, assertion := range model["p"] {
+		for index, token := range assertion.Tokens {
+			if token == fmt.Sprintf("%s_sub", ptype) {
+				subIndex = index
+				break
+			}
+		}
+		if subIndex == -1 {
+			continue
+		}
```
Comment (closetool) [inline model/model.go:225]: Can we assume `sub` always to be first, like `SortPoliciesByPriority`

File: model/model.go [hunk 3]
```diff
@@ -206,6 +208,89 @@ func (model Model) PrintModel() {
...
+		policies := assertion.Policy
+		subjectHierarchyMap, err := getSubjectHierarchyMap(model)
```
Comment (closetool) [inline model/model.go:227]: What if
```
alice -> bob
bob -> tom
tom -> alice
```

File: model/model.go [hunk 4]
```diff
@@ -206,6 +208,89 @@ func (model Model) PrintModel() {
...
+			return p1 > p2
+		})
+		for i, policy := range assertion.Policy {
+			assertion.PolicyMap[strings.Join(policy, ",")] = i
+		}
+	}
+	return nil
+}
+
+func getSubjectHierarchyMap(model Model) (map[string]int, error) {
```
Comment (closetool) [inline model/model.go:243]: directly pass in policy instead of model

File: model/model.go [hunk 5]
```diff
@@ -206,6 +208,89 @@ func (model Model) PrintModel() {
...
+	policies := model["g"]["g"].Policy
+	subjectHierarchyMap := make(map[string]int)
+	// Tree structure of role
+	policyMap := make(map[string][]string)
+	for _, policy := range policies {
+		if len(policy) < 2 {
+			return nil, errors.New("policy g expect 2 more params")
+		}
```
Comment (closetool) [inline model/model.go:251]: No need, cause casbin initialization will check this, plz check it

File: model/model.go [hunk 6]
```diff
@@ -206,6 +208,89 @@ func (model Model) PrintModel() {
...
+		child := policy[0]
+		parent := policy[1]
+		policyMap[parent] = append(policyMap[parent], child)
+		if _, ok := subjectHierarchyMap[child]; !ok {
+			subjectHierarchyMap[child] = 0
+		}
+		if _, ok := subjectHierarchyMap[parent]; !ok {
+			subjectHierarchyMap[parent] = 0
+		}
+		subjectHierarchyMap[child] = 1
```
Comment (closetool) [inline model/model.go:261]: What if there is a domain?

File: model/model.go [hunk 7]
```diff
@@ -206,6 +208,89 @@ func (model Model) PrintModel() {
...
+				if _, ok := policyMap[nodeValue]; ok {
+					for _, child := range policyMap[nodeValue] {
+						queue.PushBack(child)
+					}
+				}
+			}
+			lv++
+		}
+	}
+	return subjectHierarchyMap, nil
```
Comment (closetool) [inline model/model.go:291]: I don't get it, what you want to achieve for?

**Review by closetool — CHANGES_REQUESTED** [review]
Plz change these

File: model/model.go (hunk 3 above)
Comment (tangyang9464) [reply model/model.go:227]: I noticed that casbin itself does not support circular structure so I did't consider this unreasonable situation

File: model/model.go (hunk 7 above)
Comment (tangyang9464) [reply model/model.go:291]: This requires users to provide a reasonable tree structure.I have explained it above

File: model/model.go (hunk 5 above)
Comment (tangyang9464) [reply model/model.go:251]: https://github.com/casbin/casbin/blob/ecc6d90bd00c77235384013a5fdfe2e7aed82e66/enforcer.go#L289-L304
I found the check in ``BuildRoleLinks() ``（line 299）,but it executes after.

File: model/model.go (hunk 6 above)
Comment (tangyang9464) [reply model/model.go:261]: In this issue,the method can't support domain.I'm not sure if there is such a demand

Comment (closetool) [reply model/model.go:251]: I would recommend it

File: model/model.go (hunk 7 above)
Comment (closetool) [reply model/model.go:281]: 1. `rootArr` is not required, you could use `subjectHierarchyMap` directly.
2. `queue` could be reused, you don't need to create it in every loop.

File: model/model.go (hunk 5 above)
Comment (closetool) [reply model/model.go:241]: OK, I'll check it later.

File: model/model.go (hunk 2 above)
Comment (tangyang9464) [reply model/model.go]: In ``SortPoliciesByPriority``,I think it does’t assume that ``sub`` always to be first.
``sub`` and ``domain`` ‘s position are not fixed.Do we need to fix the position of ``sub``and ``dom`` in the model like ``g``?It seems simpler

**Comment by tangyang9464** [comment]
@closetool add domain support and "reuse".plz review again

**Review by closetool — APPROVED** [review]
LGTM

**Comment by hsluoyz** [comment]
@merajsahebdar plz review.

**Comment by merajsahebdar** [comment]
Thanks for your patience!
I played a lot around it, and find no issues. It's a great feature. 🤩🙏🏻

Just an important question, since this model, does not support multiple roles for each user (or multiple parent subjects for each child subject), can we implement something to raise an error in case of the client requests to do it? or we should just leave it up to the client?

**Comment by hsluoyz** [comment]
> Just an important question, since this model, does not support multiple roles for each user (or multiple parent subjects for each child subject), can we implement something to raise an error in case of the client requests to do it? or we should just leave it up to the client?

@tangyang9464 @closetool any comments?

**Comment by tangyang9464** [comment]
> @tangyang9464 plz add this new `subjectPriority(p.eft) || deny` feature to the docs as a new section of this page: https://casbin.org/docs/en/priority-model

I'll do it later

**Comment by tangyang9464** [comment]
> > Just an important question, since this model, does not support multiple roles for each user (or multiple parent subjects for each child subject), can we implement something to raise an error in case of the client requests to do it? or we should just leave it up to the client?
> 
> @tangyang9464 @closetool any comments?

I think it is enough to indicate in the docs

**Comment by hsluoyz** [comment]
@tangyang9464 tracked in this issue: https://github.com/casbin/casbin-website/issues/252

**Comment by hsluoyz** [comment]
@TraMiu create issue at jcasbin repo

**Pull request merged**